File: app/src/main/java/org/nitri/opentopo/nearby/api/NearbyDatabase.java
package org.nitri.opentopo.nearby.api;

import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import android.content.Context;

import org.nitri.opentopo.nearby.da.NearbyDao;
import org.nitri.opentopo.nearby.entity.NearbyItem;

@Database(entities = {NearbyItem.class}, version = 1, exportSchema = false)
public abstract class NearbyDatabase extends RoomDatabase {

    private static NearbyDatabase instance;

    public abstract NearbyDao nearbyDao();

    public static synchronized NearbyDatabase getDatabase(Context context) {
        if (instance == null) {
            instance =
                    Room.databaseBuilder(context.getApplicationContext(), NearbyDatabase.class, "nearby-database")
                            .build();
        }
        return instance;
    }

}

File: app/src/main/java/org/nitri/opentopo/nearby/api/NearbyDatabase.kt
package org.nitri.opentopo.nearby.api

import android.content.Context
import androidx.room.Database
import androidx.room.Room.databaseBuilder
import androidx.room.RoomDatabase
import org.nitri.opentopo.nearby.da.NearbyDao
import org.nitri.opentopo.nearby.entity.NearbyItem

@Database(entities = [NearbyItem::class], version = 1, exportSchema = false)
abstract class NearbyDatabase : RoomDatabase() {
    abstract fun nearbyDao(): NearbyDao?

    companion object {
        @Volatile
        private var instance: NearbyDatabase? = null

        fun getDatabase(context: Context): NearbyDatabase {
            return instance ?: synchronized(this) {
                // Double-check in case 'instance' was initialized while this thread was blocked
                instance ?: buildDatabase(context).also { instance = it }
            }
        }

        private fun buildDatabase(context: Context): NearbyDatabase {
            return databaseBuilder(
                context.applicationContext,
                NearbyDatabase::class.java,
                "nearby-database"
            ).build()
        }
    }
}
